qiskit/backends/ibmq/ibmqjob.py

```diff
@@ -186,10 +192,55 @@ def result(self, timeout=None, wait=5):
         except ApiError as api_err:
             raise JobError(str(api_err))
 
-        if self._is_device and self.status() == JobStatus.DONE:
-            _reorder_bits(job_data)
-
         # Build the Result.
+        experiment_results = []
+        if 'result' in job_data:  # qobj job
+            job_result = job_data['result']
+            for resultobj in job_result['results']:
+                qobj_exp_result_args = [resultobj.get(arg) for arg in
+                                        QobjExperimentResult.REQUIRED_ARGS]
+                qobj_exp_result_kwargs = {key: value for (key, value) in
+                                          resultobj.items() if key not in
+                                          QobjExperimentResult.REQUIRED_ARGS}
+
+                qobj_exp_result = QobjExperimentResult(*qobj_exp_result_args,
+                                                       **qobj_exp_result_kwargs)
+                experiment_results.append(qobj_exp_result)
+            qobj_result_args = [job_result.get(arg) for arg in
+                                QobjResult.REQUIRED_ARGS]
+            qobj_result_kwargs = {key: value for (key, value) in
+                                  job_result.items() if key not in
+                                  QobjResult.REQUIRED_ARGS}
+            qobj_result = QobjResult(*qobj_result_args, **qobj_result_kwargs)
+            # replace job_result list of dict with list of Qobj ExperimentResult
+            qobj_result.results = experiment_results
+            return Result(qobj_result)
+        elif 'qasms' in job_data:  # old-style qasms job
+            if self._is_device and self.status() == JobStatus.DONE:
```

Comment:
Non-blocking: here is something we did not catch before. If, while asking for a result, it turns out the job finishes in the `ERROR` state, we should raise and inform the user she can not ask for a result in this state. So, right after [retrieving `job_data`](https://github.com/Qiskit/qiskit-terra/pull/713/files#diff-1601b1931db3e408a0c7973d5605ff4aR191), we should add a checking for the `ERROR` state and raise accordingly.

Even if not part of the scope of this PR, would you mind to fix it now or file an issue instead, please?

@atilag do you mind to take a look at this? does it sound reasonable?

Reply:
So if user code calls `job.result()` without previously checking the status... which is a valid use-case, **throwing in case of an error response from the server is fine**, as we are keeping `Result` object as the representation of a valid result, so there's no concept of error result. Throwing will happen sooner or later anyway.